test/preparser.js

```diff
@@ -37,7 +38,63 @@ describe('preparser', function () {
       }
     });
 
-    it.skip('should convert the same variable twice', function () {
+    it('concatenate variables with no space', function () {
+      if (windows) {
+        cash('echo %PATH%%PATH%').should.equal(`${path}${path}\n`);
+      } else {
+        cash('echo $PATH$PATH').should.equal(`${path}${path}\n`);
+      }
+    });
+
+    it('display variables with one or more spaces in between', function () {
+      if (windows) {
+        cash('echo %PATH% %PATH%').should.equal(`${path} ${path}\n`);
+      } else {
+        cash('echo $PATH $PATH').should.equal(`${path} ${path}\n`);
+      }
+    });
+
+    it('display variable references from inside double quotes', function () {
+      if (windows) {
+        cash('echo "%PATH%   %PATH%"').should.equal(`${path}   ${path}\n`);
+      } else {
+        cash('echo "$PATH   $PATH"').should.equal(`${path}   ${path}\n`);
+      }
+    });
+
+    it('separate variable references from strings with slash or dot', function () {
+      if (windows) {
+        cash('echo foo/%PATH%/%PATH%.bar').should.equal(`foo/${path}/${path}.bar\n`);
+      } else {
+        cash('echo foo/$PATH/$PATH.bar').should.equal(`foo/${path}/${path}.bar\n`);
+      }
+    });
+
+    it('non-existent variables default to the empty string for unix', function () {
+      if (!windows) {
+        cash('echo $FAKE_ENV_VAR').should.equal('\n');
+      }
+    });
+
+    it('underscores can be in environmental variable names', function () {
+      if (nvmDir) {
+        if (windows) {
+          cash('echo %NVM_DIR%').should.equal(`${nvmDir}\n`);
+        } else {
+          cash('echo $NVM_DIR').should.equal(`${nvmDir}\n`);
+        }
+      }
+    });
+
+    it('append variables and strings', function () {
+      if (windows) {
+        cash('echo foo%PATH%bar').should.equal(`foo${path}bar\n`);
+      } else {
+        cash('echo foo${PATH}bar').should.equal(`foo${path}bar\n`);
+      }
+    });
+
```

Comment:
Can we add one (at least for Windows) that deals with really tricky strings such as:

`echo hi%invalid%computername%username%username%%%`

Given `%computername%` is `dc-comp` and `%username%` is `dc`, the output should be:

`hi%invaliddc-compusernamedc%%`

This should work already, but it would be good to have a test for it.